templates/common/on-prem/files/NetworkManager-resolv-prepender.yaml

```diff
@@ -20,89 +20,116 @@ contents:
+        case "$STATUS" in
+            up|dhcp4-change|dhcp6-change)
+            >&2 echo "NM resolv-prepender triggered by ${IFACE} ${STATUS}."
+
+            # In DHCP connections, the resolv.conf content may be late, thus we wait for nameservers
             while ! grep nameserver /var/run/NetworkManager/resolv.conf; do
                 >&2 echo  "NM resolv-prepender: NM resolv.conf still empty of nameserver"
                 sleep 0.5
             done
-    EOF
-        # Ensure resolv.conf exists and contains nameservers before we try to run podman
-        if [[ ! -e /etc/resolv.conf ]] || ! grep -q nameserver /etc/resolv.conf; then
-            cp /var/run/NetworkManager/resolv.conf /etc/resolv.conf
-        fi
 
-        # Pull container image outside of the timeout to workaround possible
-        # image corruption
-        # https://github.com/containers/podman/issues/14003
-        /usr/bin/podman pull --authfile /var/lib/kubelet/config.json {{ .Images.baremetalRuntimeCfgImage }}
+            # Ensure resolv.conf exists and contains nameservers before we try to run podman
+            if [[ ! -e /etc/resolv.conf ]] || ! grep -q nameserver /etc/resolv.conf; then
+                cp /var/run/NetworkManager/resolv.conf /etc/resolv.conf
+            fi
 
-        NAMESERVER_IP=$(timeout 20s /usr/bin/podman run --rm \
-            --authfile /var/lib/kubelet/config.json \
-            --net=host \
-            {{ .Images.baremetalRuntimeCfgImage }} \
-            node-ip \
-            show \
-            --retry-on-failure \
-            {{range onPremPlatformAPIServerInternalIPs . }}"{{.}}" {{end}} \
-            {{range onPremPlatformIngressIPs . }}"{{.}}" {{end}})
-        DOMAINS="${IP4_DOMAINS} ${IP6_DOMAINS} {{.DNS.Spec.BaseDomain}}"
-        if [[ -n "$NAMESERVER_IP" ]]; then
-            if systemctl -q is-enabled systemd-resolved; then
-                >&2 echo "NM resolv-prepender: Setting up systemd-resolved for OKD domain and local IP"
-                if [[ ! -f /etc/systemd/resolved.conf.d/60-kni.conf ]]; then
+            # Pull container image outside of the timeout in a background process to workaround possible
+            # image corruption
+            # https://github.com/containers/podman/issues/14003
+            if ! /usr/bin/podman image exists {{ .Images.baremetalRuntimeCfgImage }}; then
+                # Multiple podman pulls may occur in parallel wasting bandwidth therefore this
+                # needs to be removed when podman image corruption issue is resolved.
+                # Send to background to avoid SIGTERM from timeout which may cause image corruption issue during image pull
+                pull_baremetal_runtime_cfg_image &
+                # Wait until image is available
+                >&2 echo "NM resolv-prepender: Waiting for baremetal runtime cfg image to be available"
+                until /usr/bin/podman image exists {{ .Images.baremetalRuntimeCfgImage }}; do
```

Comment:
What happens if the image pull takes longer than the 30s timeout granted to run the `resolv_prepender` function?

I could be missing something obvious, but it looks to me that the pull will be killed because the parent process dies. Check this small script:
```bash
function resolv_prepender {
        longer_than_timeout &
        while :; do
                sleep 1
        done
        echo unreachable
}

function longer_than_timeout {
        for i in 1 2 3; do
                echo $i
                sleep 1
        done
        echo done
}

export -f resolv_prepender longer_than_timeout
timeout 2s bash -c resolv_prepender
```

```
/tmp
❯ bash ./test.sh
1
2

/tmp
❯ echo $?
124
```
If that's really the case, we shouldn't be bothering with running the pull as a separate command. Or alternatively, we should pull the image before we run the dispatcher script. Is that even possible?

Reply:
Duh. Thank you @mandre. Fixed it. Appreciate your eyes here.

Reply:
Ive asked a few folks and no one has come up any ideas how to do this after network manager is up but before dispatcher scripts are invoked. If you have any ideas, please let me know.

Reply:
Regarding whether this child process will continue after the dispatcher script is run:
```
Dispatcher scripts are run one at a time, but asynchronously from the main NetworkManager process, and will be killed if they run for too long. If your script might take arbitrarily long to complete, you should spawn a child process and have the parent return immediately. 
```
https://networkmanager.dev/docs/api/latest/NetworkManager-dispatcher.html

I tested that the child process will continue following return of the dispatcher script.
We dont return immediately, but if we respect NM timeout, we should be ok.